<a href="https://colab.research.google.com/github/grunobuide/portfolio/blob/main/Insurance_BMI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Medical Insurance Costs Project

Datascience exercise in codecademy - reviewing things that I already know and
 updating my python skills.


##Goal

Finding out if we can use linear regression to predict the charges of medical insurance.

##To do

- import the data
- save in python friendly format
- use a linear regression model (y = mx + b), where  y means charges and x is the variable of interest.
- make a gridsearch to find best values for m and b given the interval.
- split the dataset into training and testing
- find the average error of the model.



##Script


In [ ]:
import csv
from sklearn import model_selection as m_s
#cutting corners instead of implementing a random function to split de dataset in training/test

with open('/insurance.csv') as csv_file:
  #opening the file and then creating the datapoints x is bmi, y is charges
  #just by changing the name of the first value, i can test different variables
  #if they are numerical - 'children','age', and 'bmi'.
  csv_dict = csv.DictReader(csv_file)
  datapoints = [(float(x['children']),float(x['charges'])) for x in csv_dict]



#datapoints
#train_data

In [ ]:
#normalize the dataset - shortcoming: if i find a new observation which is smaller
#or larger than the min/max in the dataset, my normalization is caput.
#but this technique is interesting to use if you find the dataset representative
#of the population, so for this basic exercise i've considered it good enough.

def normalize(dataset):
  min_x = min([x[0] for x in dataset])
  min_y = min([x[1] for x in dataset])
  max_x = max([x[0] for x in dataset])
  max_y = max([x[1] for x in dataset])
  output = []
  for x,y in dataset:
    out_x = (x-min_x)/(max_x-min_x)
    out_y = (y-min_y)/(max_y-min_y)
    output.append((out_x,out_y))
  return output


n_datapoints = normalize(datapoints)
#normal_test = [(normalize(x[0]),normalize(x[1])) for x in test_data]
#normal_train
test_data, train_data = m_s.train_test_split(n_datapoints, test_size=0.2)

In [ ]:
#functions to perform linear regression

def get_y (m,b,x):
  return m*x + b



def error(m,b,points):
  total_error = 0
  for point in points:
    x = float(point[0])
    y = float(point[1])
    total_error += abs(y - get_y(m,b,x))
  return total_error/float(len(points))

#generate possible ms and bs

possible_ms = [x/100 for x in range(-100,101)]
possible_bs = [x/100 for x in range(-100,101)]




#for each m and b, get the error, save the min error with the best m and b

minimal_error = float('inf')
best_m = 0
best_b = 0

for m in possible_ms:
  for b in possible_bs:
    current_error = error(m,b,train_data)
    if current_error < minimal_error:
      minimal_error = current_error
      best_m = m
      best_b = b

print(best_m, best_b, minimal_error)


-0.02 0.13 0.13290265186866057


In [ ]:
error(best_m,best_b,test_data)

0.133809620500867

##Análise

I've got a linear regression model with a dataset specific normalization of the values.

The model has an average error of 13% in predicting the charge based on the bmi.

Trying with age: the average error is 10%

With n of children: 13%